p2p/libp2p/module.go

```diff
@@ -0,0 +1,344 @@
+	bus                   modules.Bus
+	cfg                   *configs.P2PConfig
+	addrBookProvider      providers.AddrBookProvider
+	currentHeightProvider providers.CurrentHeightProvider
+	identity              libp2p.Option
+	listenAddrs           libp2p.Option
+	host                  host.Host
+	pubsub                *pubsub.PubSub
+	topic                 *pubsub.Topic
+	subscription          *pubsub.Subscription
+	network               typesP2P.Network
+}
+
+// NewReadStreamDeadline returns a future deadline
+// based on the read stream timeout duration.
+func NewReadStreamDeadline() time.Time {
+	return time.Now().Add(readStreamTimeoutDuration)
+}
+
+var (
+	// TODO: consider security exposure of and "safe minimum" for timeout.
+	// TODO: parameterize and expose via config.
+	// ReadStreamTimeout is the duration to wait for a read operation on a
+	// stream to complete, after which the stream is closed ("timed out").
+	readStreamTimeoutDuration = time.Second * 10
+	// ErrModule wraps errors which occur within the libp2pModule implementation.
+	ErrModule = common.NewErrFactory("LibP2P module error")
+	// ErrCloseStream wraps errors which occur when attempting to close a peer stream.
+	ErrCloseStream = common.NewErrFactory("an error occurred while closing peer stream")
+)
+
+func Create(bus modules.Bus) (modules.Module, error) {
+	return CreateWithProviders(
+		bus,
+		persistence.NewPersistenceAddrBookProvider(bus),
+		bus.GetConsensusModule(),
+	)
+}
+
+func CreateWithProviders(bus modules.Bus, addrBookProvider addrbook_provider.AddrBookProvider, currentHeightProvider providers.CurrentHeightProvider) (modules.Module, error) {
+	return new(libp2pModule).CreateWithProviders(bus, addrBookProvider, currentHeightProvider)
+}
+
+func (mod *libp2pModule) GetModuleName() string {
+	// TODO: double check if this should change.
+	return modules.P2PModuleName
+}
+
+func (mod *libp2pModule) Create(bus modules.Bus) (modules.Module, error) {
+	return Create(bus)
+}
+
+func (mod *libp2pModule) CreateWithProviders(bus modules.Bus, addrBookProvider addrbook_provider.AddrBookProvider, currentHeightProvider providers.CurrentHeightProvider) (modules.Module, error) {
+	*mod = libp2pModule{
+		addrBookProvider:      addrBookProvider,
+		currentHeightProvider: currentHeightProvider,
+	}
+
+	if err := bus.RegisterModule(mod); err != nil {
+		return nil, ErrModule("unable to register module", err)
+	}
+
+	log.Println("Creating libp2p-backed network module")
+
+	mod.cfg = bus.GetRuntimeMgr().GetConfig().P2P
+
+	// TODO: support RainTree network
+	if mod.cfg.UseRainTree {
+		return nil, ErrModule("raintree is not yet compatible with libp2p", nil)
+	}
+
+	// TODO(future): investigate any unnecessary
+	// key exposure / duplication in memory
+	secretKey, err := identity.NewLibP2PPrivateKey(mod.cfg.PrivateKey)
+	if err != nil {
+		// TODO: wrap error
+		return nil, err
+	}
+
+	mod.identity = libp2p.Identity(secretKey)
+	// TODO: where is the hostname specified?
```

Comment:
I think it would make sense to allow node operators to specify the interface on which to listen. Does such a thing already exists and I just missed it?

Reply:
Search for `serviceUrlFormat`: `serviceUrlFormat  = "node%d.consensus:8080"`

consensus_port is set to `8080` in configs, but I just realize it's a bug that we don't use it

Reply:
I did notice the port in the consensus config, thanks! I was referring specifically to being able to specify a hostname (i.e. `<hostname>:<port>`), in case we intend to support DNS resolution in production. Somewhat orthogonally, in cases where a node operator has multiple network interfaces (excl. loopback for this discussion), I think it would be prudent to allow a node operator to choose exactly which to expose. I didn't see any such degree of freedom in what I read so far of the current startup / config process. Perhaps this is related to some of the post-genesis address book work that I've seen references to?

It's not clear to me how the `serviceUrlFormat` piece connects as I only see references to test code when searching, could you elaborate a bit more?

Reply:
Discussed offline but let me know if we need to reiterate here.

1. The `p2pConfig` exposes a port (e.g. 8080).
2. This port is where the node listens for consensus and p2p-specific events
3. Implicitly, we append the hostname to the `8080` port (e.g. `node1.consensus`) so we can send messages to another node based on their validator number
4. The `serviceUrl` also mimics (not enforced) what `(3):(1)` is based on the points above
5. The `serviceUrl` is defined via `generic_param` in `genesis.json`

NOTE: This is not done by design; this just happened and should change.

@okdas Have you put any thought into how these configs should be done?